tensorflow_probability/python/mcmc/replica_exchange_mc.py

```diff
@@ -0,0 +1,426 @@
+        self._seed_stream = distributions_util.gen_new_seed(
+            self._seed_stream, salt='replica_exchange_one_step')
+        log_uniform = tf.log(tf.random_uniform(
+            shape=tf.shape(ratio),
+            dtype=ratio.dtype.base_dtype,
+            seed=self._seed_stream))
+        exchange = log_uniform < ratio
+        exchange_op = tf.sparse_to_dense(
+            [exchange_proposed[i, 0], exchange_proposed[i, 1]],
+            [self.num_replica],
+            [next_replica_idx[exchange_proposed[i, 1]] -
+             next_replica_idx[exchange_proposed[i, 0]],
+             next_replica_idx[exchange_proposed[i, 0]] -
+             next_replica_idx[exchange_proposed[i, 1]]])
+        next_replica_idx = tf.cond(exchange,
+                                   lambda: next_replica_idx + exchange_op,
+                                   lambda: next_replica_idx)
+        return [i + 1, next_replica_idx]
+
+      next_replica_idx = tf.while_loop(
+          cond, body, loop_vars=[i, next_replica_idx])[1]
+
+      next_replica_states = []
+      next_replica_results = []
+      for i in range(self.num_replica):
+        next_replica_states.append(
+            tf.case({tf.equal(next_replica_idx[i], j):
+                    _stateful_lambda(sampled_replica_states[j])
+                    for j in range(self.num_replica)}, exclusive=True))
+        next_replica_results.append(
+            tf.case({tf.equal(next_replica_idx[i], j):
+                     _stateful_lambda(sampled_replica_results[j])
+                     for j in range(self.num_replica)}, exclusive=True))
+
+      next_state = tf.identity(next_replica_states[0])
+      kernel_results = ReplicaExchangeMCKernelResults(
+          replica_results=next_replica_results,
+          replica_states=next_replica_states,
+          next_replica_idx=next_replica_idx,
+          exchange_proposed=exchange_proposed,
+          exchange_proposed_n=exchange_proposed_n,
+      )
+
+      return next_state, kernel_results
+
+  def bootstrap_results(self, init_state):
+    """Returns an object with the same type as returned by `one_step`.
+
+    Args:
+      init_state: `Tensor` or Python `list` of `Tensor`s representing the
+        a state(s) of the Markov chain(s).
+
+    Returns:
+      kernel_results: A (possibly nested) `tuple`, `namedtuple` or `list` of
+        `Tensor`s representing internal calculations made within this function.
+        This inculdes replica states.
+    """
+    with tf.name_scope(
+        name=mcmc_util.make_name(self.name, 'remc', 'bootstrap_results'),
+        values=[init_state]):
+      replica_results = [self.replica_kernels[i].bootstrap_results(init_state)
+                         for i in range(self.num_replica)]
+
+      init_state_parts = (list(init_state)
+                          if mcmc_util.is_list_like(init_state)
+                          else [init_state])
+      replica_states = [[tf.Variable(s) if isinstance(s, tf.Variable)
```

Comment:
We try to avoid using `tf.Variable` or `tf.get_variable` in any of our API code.  I dont think you actually need to create a variable here. Ie, the user provided `target_log_prob_fn` should not presume its input is a `tf.Variable` instance.

Typically users would do something like:

    state = tf.get_variable("state", shape=[], dtype=np.float32)
    state = tf.assign(state, hmc.one_step(state, hmc.bootstrap_results(state))[0])

Ie, all handling of variables should happen in user code, not API code.